.github/workflows/integrate.yaml

```diff
@@ -4,92 +4,121 @@ name: CI
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: ${{ secrets.AWS_REGION }}
+      - name: Start EC2 runner
+        id: start-ec2-runner
+        uses: machulav/ec2-github-runner@v2
+        with:
+          mode: start
+          github-token: ${{ secrets.GH_PERSONAL_ACCESS_TOKEN }}
+          ec2-image-id: ${{ vars.AWS_EC2_IMAGE_ID }}
+          ec2-instance-type: ${{ vars.AWS_EC2_INSTANCE_TYPE }}
+          subnet-id: ${{ vars.AWS_DEFAULT_SUBNET_ID }}
+          security-group-id: ${{ vars.AWS_SECURITY_GROUP_ID }}
   lint:
-    name: Lint
-    runs-on: ubuntu-20.04
-    strategy:
-      fail-fast: false
-      matrix:
-        charm:
-        - server
+    name: Run lint
+    needs: start-runner
+    runs-on: ${{ needs.start-runner.outputs.label }}
     steps:
-    - uses: actions/checkout@v2
-    - run: python3 -m pip install tox
-    - run: tox -e ${{ matrix.charm }}-lint
-
-  unit:
-    name: Unit tests
-    runs-on: ubuntu-20.04
-    strategy:
-      fail-fast: false
-      matrix:
-        charm:
-        - server
+    - name: Check out repo
+      uses: actions/checkout@v2
+    - name: Install dependencies
+      run: sudo apt-get install -y python3-pip tox
+    - name: Lint code
+      run: tox -e lint
+  unit-tests:
+    name: Run unit tests
+    needs: [start-runner, lint]  
+    runs-on: ${{ needs.start-runner.outputs.label }}
     steps:
-    - uses: actions/checkout@v2
-    - run: python3 -m pip install tox
-    - run: tox -e ${{ matrix.charm }}-unit
+    - name: Check out code
+      uses: actions/checkout@v2
+
+    - name: Install dependencies
+      run: sudo apt-get install -y python3-pip tox
 
-  integration:
-    name: Integration tests (microk8s)
-    runs-on: ubuntu-20.04
-    strategy:
-      fail-fast: false
-      matrix:
-        charm:
-        - server
+    - name: Run unit tests
+      run: tox -e unit
+  integration-tests:
+    name: Run integration tests
+    needs: [start-runner, lint, unit-tests]
+    runs-on: ${{ needs.start-runner.outputs.label }} 
     steps:
-    - uses: actions/checkout@v2
-    - name: Setup operator environment
-      # TODO: Unpin this when https://github.com/charmed-kubernetes/actions-operator/pull/46 is merged
-      uses: claudiubelu/actions-operator@18ebf92ae3043bd3dd15238e5d9b662d7ba08daf
-      with:
-        provider: microk8s
-        channel: 1.24/stable
-        # Pinned until this bug is resolved: https://bugs.launchpad.net/juju/+bug/1992833
-        bootstrap-options: "--agent-version=2.9.34"
-        microk8s-addons: "dns storage rbac metallb:10.64.140.43-10.64.140.49"
-    - name: Install firefox
-      run: sudo apt install -y firefox
+    - name: Check out repo
+      uses: actions/checkout@v2
 
+    - name: Setup operator environment
+      # The runner will run the code as root. There is problem when using juju under rood so here is small workaround to exec as ubuntu user
+      run: |
+        chmod 777 .
```

Comment:
Can we limit permissions or does it strictly have to be `777`?

Reply:
no User is also creating .tox folder so it needs to be 777 (I tried but not enough otherwise)